src/Transients.cc

```diff
@@ -336,13 +336,23 @@ Transients::readers(const StoreEntry &e) const
 void
 Transients::markForUnlink(StoreEntry &e)
 {
-    unlink(e);
+    assert(e.key);
+    e.hasTransients() ? unlink(e) :
+        unlinkByKeyIfFound(reinterpret_cast<const cache_key*>(e.key));
+}
+
+void
+Transients::unlinkByKeyIfFound(const cache_key *key)
+{
+    // Controller ensures that this worker has no StoreEntry to abandon() here.
+    if (map)
+        map->freeEntryByKey(key);
 }
 
 void
 Transients::unlink(StoreEntry &e)
 {
-    if (e.mem_obj && e.mem_obj->xitTable.io == MemObject::ioWriting)
+    if (e.mem_obj)
```

Comment:
> Otherwise, unlink() has to check hasTransients() and call markForUnlink() if there are none (approach 2).

After these changes I noticed that unlink() and markForUnlink() became identical. So I made unlink() private(which now only calls markForUnlink()) and use markForUnlink() in that single place instead.

Reply:
Done at a40bbf6.